### PR TITLE
feat: add keys to organization provision daemons

### LINKED ISSUE #14439: API: Organization Provisioner Page
The tasks required for a functional page are:
- [ ] Ensure PSK provisioners can only connect to the default org
- [x] Define what "active" means for a provisioner and compose a way to return only active provisioners
- [x] Ensure we can compare provisioner protocol version (for out of date warnings)
- [x] Enumerate types of warnings possible so frontend can design
- [x] Connect provisioner keys and provisioner connections - Add reference `key` field to provisioners table
- [x] Return list of provisioners to frontend grouped by authentication type: provisioner key, PSK, and built in 
  - Provisioner key fields: name, tags, created at, last used
  - Provisioner fields: name, scope, tags (if applicable), and version

### PR DESCRIPTION
**Pull request by f0ssel**
Closes https://github.com/coder/coder/issues/14439

What is changes:
- Adds `KeyID` field to provisioner daemons
- Hardcodes 3 static provisioner daemon keys in the database for the default org - `built-in`, `user-auth`, and `psk`
- Uses static provisioner keys when inserting new provisioner daemons that are using those auth methods
- Adds new endpoint that lists provisioner keys each with an embedded list of associated daemons
- Fixes some logic with key auth

I'd like to add more that test deeper functionality on the get provisioner daemons endpoint. I'll do this in a followup so this can be merged and unblock UI work.

### CODE REVIEW

File: codersdk/provisionerdaemons.go [hunk 1]
```diff
@@ -282,6 +283,12 @@ type ProvisionerKey struct {
 	// HashedSecret - never include the access token in the API response
 }
 
+const (
+	ProvisionerKeyIDBuiltIn  = "11111111-1111-1111-1111-111111111111"
+	ProvisionerKeyIDUserAuth = "22222222-2222-2222-2222-222222222222"
+	ProvisionerKeyIDPSK      = "33333333-3333-3333-3333-333333333333"
+)
+
```
Comment (f0ssel) [inline codersdk/provisionerdaemons.go:291]: These will probably need to be hardcoded on the frontend as well, since this doesn't seem to get generated.

File: coderd/database/db2sdk/db2sdk.go [hunk 2]
```diff
@@ -533,13 +533,38 @@ func ProvisionerDaemon(dbDaemon database.ProvisionerDaemon) codersdk.Provisioner
...
 	return result
 }
 
+func RecentProvisionerDaemons(now time.Time, staleInterval time.Duration, daemons []database.ProvisionerDaemon) []codersdk.ProvisionerDaemon {
+	results := []codersdk.ProvisionerDaemon{}
+
+	// Ensure stable order for display and for tests
+	sort.Slice(daemons, func(i, j int) bool {
+		return daemons[i].Name < daemons[j].Name
+	})
```
Comment (Emyrk) [inline coderd/database/db2sdk/db2sdk.go:550]: Maybe sort `results` instead to not modify the inputted slice. Since sort does it in place, just sort the new slice.

File: enterprise/coderd/provisionerkeys.go [hunk 3]
```diff
@@ -54,6 +55,20 @@ func (api *API) postProvisionerKey(rw http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	reserved := []string{"built-in", "psk", "user-auth"}
```
Comment (Emyrk) [inline enterprise/coderd/provisionerkeys.go:58]: Do you think there is a place we can make this constants? I saw this also used in `dbmem` and some unit tests.

Not a hugel deal if it feels like a pain.

File: coderd/database/queries/provisionerkeys.sql [hunk 4]
```diff
@@ -43,7 +43,13 @@ SELECT
...
     provisioner_keys
 WHERE
-    organization_id = $1;
+    organization_id = $1
+AND
+    id != '11111111-1111-1111-1111-111111111111'::uuid
+AND 
+    id != '22222222-2222-2222-2222-222222222222'::uuid
+AND 
+    id != '33333333-3333-3333-3333-333333333333'::uuid;
```
Comment (Emyrk) [inline coderd/database/queries/provisionerkeys.sql:52]: Drop some comments on why these are filtered out

File: coderd/database/migrations/000250_built_in_psk_provisioner_keys.up.sql [hunk 5]
```diff
@@ -0,0 +1,5 @@
+INSERT INTO provisioner_keys (id, created_at, organization_id, name, hashed_secret, tags) VALUES ('11111111-1111-1111-1111-111111111111'::uuid, NOW(), (SELECT id FROM organizations WHERE is_default = true), 'built-in', ''::bytea, '{}');
+INSERT INTO provisioner_keys (id, created_at, organization_id, name, hashed_secret, tags) VALUES ('22222222-2222-2222-2222-222222222222'::uuid, NOW(), (SELECT id FROM organizations WHERE is_default = true), 'user-auth', ''::bytea, '{}');
+INSERT INTO provisioner_keys (id, created_at, organization_id, name, hashed_secret, tags) VALUES ('33333333-3333-3333-3333-333333333333'::uuid, NOW(), (SELECT id FROM organizations WHERE is_default = true), 'psk', ''::bytea, '{}');
+
+ALTER TABLE provisioner_daemons ADD COLUMN key_id UUID REFERENCES provisioner_keys(id) ON DELETE CASCADE DEFAULT '11111111-1111-1111-1111-111111111111'::uuid NOT NULL;
```
Comment (Emyrk) [inline coderd/database/migrations/000250_built_in_psk_provisioner_keys.up.sql:5]: We have external provisioners running today on some customer deployments. Will this be ok if they are assigned to the built in?

Should we assign the existing rows to the `psk` at `33333333-3333-3333-3333-33333333333`, and have the startup code fix the built in rows?

Comment (f0ssel) [reply coderd/database/migrations/000250_built_in_psk_provisioner_keys.up.sql:5]: So I did try to consider the outputs here and picked this one since it felt like the least impact, but let me know if you feel otherwise:
- There's no data in the DB that lets us know if it's a built-in or PSK today
- External provisioners will need to reconnect after a coderd upgrade anyways, and on reconnection they will be assigned the correct `333...` id. 
- On the average deployment we would most likely see more built-ins than external, so if we have to pick one, built-in is probably less incorrect
- I considered allowing null and not back porting the reference but that felt extra messy in the resulting queries and handler logic.

So basically with this migration the external provisioners will be incorrectly labeled as built-in until the provisioners reconnect. Because we only show recent provisioners the ones that do not connect again (and stay incorrectly labeled) will only be shown for a short while before being removed from the list we show users.

Comment (f0ssel) [reply coderd/database/migrations/000250_built_in_psk_provisioner_keys.up.sql:5]: Also worth noting that right now this key reference is only for grouping in read operations, no mutation logic is performed based on this value today.

**Review by Emyrk — APPROVED** [review]
The reconnecting external provisioners makes sense :+1:

**Comment by aslilac** [comment]
I feel very weird about hardcoding certain UUIDs as a "type" indicator

Also, I think I totally spaced before on this, we'll want the key name on the front-end as well

**Comment by f0ssel** [comment]
> I feel very weird about hardcoding certain UUIDs as a "type" indicator
> 
> Also, I think I totally spaced before on this, we'll want the key name on the front-end as well

Yeah I get the hesitation, I'll try to layout my thought process that led me here:
- If we add a enum `type` field, we'll still need to connect the provisioner with the key used, so we'd need 2 fields `type` and `key_id` anyways. 
- If we don't hardcode the key ids to essentially enum types, we'll need to support null references in the `key_id` field which I think we've successfully avoid in our DB schema so far, so I'd like to keep it that way. 
- I pictured the frontend fetching both keys and provisioners in parallel and joining them, so that's where the key data like `name` would come from.
- I didn't want to embed the key in each provisioner, too much duplicated data
- I didn't embed the provisioners inside each key, but we could do this and have the data mirror the page layout structure. I remember us talking and saying we _didn't_ want this for some reason so I didn't do that. 

I'm down to change the structure, so even if we use special IDs on the backend we can shim it into a "fake key" for the frontend response if that's needed. I think listing each key with a list of provisioners embedded in each key may be the best format IMO.

**Comment by aslilac** [comment]
also this is kind of a minor thought, but I might suggest

```go
const (
	ProvisionerKeyIDBuiltIn  = "00000000-0000-0000-0000-000000000001"
	ProvisionerKeyIDUserAuth = "00000000-0000-0000-0000-000000000002"
	ProvisionerKeyIDPSK      = "00000000-0000-0000-0000-000000000003"
)
```

I don't necessarily expect the number of these to balloon, but the current way gives us a maximum of 15 variants, this would allow for billions, just by using the last segment of the ID. plus it's a bit less "busy" visually.

**Comment by Emyrk** [comment]
I am not opposed to the special IDs. I think it is better than assigning them actual UUIDs. As for 0s or 1s or w/e, I also took a few moments considering what the best static UUIDs should be.

I was looking at HexWords to see if anything would be fun
https://www.nsftools.com/tips/HexWords.htm

**Comment by f0ssel** [comment]
I've added a new endpoint that returns a list of keys each with their own embedded list of associated daemons. I've also changed the UUIDs to match what @aslilac suggested. I opted for a new endpoint so we don't break dev, and in a followup PR we can remove the old list endpoint.

File: coderd/coderd.go [hunk 6]
```diff
@@ -1502,6 +1502,11 @@ func (api *API) CreateInMemoryTaggedProvisionerDaemon(dialCtx context.Context, n
 		dbTypes = append(dbTypes, database.ProvisionerType(tp))
 	}
 
+	keyID, err := uuid.Parse(string(codersdk.ProvisionerKeyIDBuiltIn))
+	if err != nil {
+		return nil, xerrors.Errorf("failed to parse built-in provisioner key ID: %w", err)
+	}
```
Comment (aslilac) [inline coderd/coderd.go:1508]:
```suggestion
	keyID := uuid.MustParse(codersdk.ProvisionerKeyIDBuiltIn)
```

1. a built-in constant we control feels like a good target for `uuid.MustParse`
2. these are already just basic strings, yeah?

```go
const (
	ProvisionerKeyIDBuiltIn  = "00000000-0000-0000-0000-000000000001"
	ProvisionerKeyIDUserAuth = "00000000-0000-0000-0000-000000000002"
	ProvisionerKeyIDPSK      = "00000000-0000-0000-0000-000000000003"
)
```

File: enterprise/coderd/provisionerdaemons.go [hunk 7]
```diff
@@ -71,7 +71,10 @@ func (api *API) provisionerDaemons(rw http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	httpapi.Write(ctx, rw, http.StatusOK, db2sdk.List(daemons, db2sdk.ProvisionerDaemon))
+	// provisionerdserver.DefaultHeartbeatInterval*3 matches the healthcheck report staleInterval.
+	recentDaemons := db2sdk.RecentProvisionerDaemons(time.Now(), provisionerdserver.DefaultHeartbeatInterval*3, daemons)
```
Comment (aslilac) [inline enterprise/coderd/provisionerdaemons.go:75]: this'll prevent the frontend from warning about non-recent provisioners. I could see this being useful as a query/filter parameter but I'm not sure it's the right default.

File: enterprise/coderd/provisionerkeys.go [hunk 8]
```diff
@@ -98,6 +115,50 @@ func (api *API) provisionerKeys(rw http.ResponseWriter, r *http.Request) {
...
+
+	daemons, err := api.Database.GetProvisionerDaemonsByOrganization(ctx, organization.ID)
+	if err != nil {
+		httpapi.InternalServerError(rw, err)
+		return
+	}
+	// provisionerdserver.DefaultHeartbeatInterval*3 matches the healthcheck report staleInterval.
+	recentDaemons := db2sdk.RecentProvisionerDaemons(time.Now(), provisionerdserver.DefaultHeartbeatInterval*3, daemons)
+
+	pkDaemons := make([]codersdk.ProvisionerKeyDaemons, 0, len(sdkKeys))
```
Comment (aslilac) [inline enterprise/coderd/provisionerkeys.go:145]: if there's more than one provisioner per key on average, this'll reallocate. a safer upper-bound would be `len(recentDaemons)`, but there is the possibility that it could overshoot by a bit. I'm fine with either, just making sure the tradeoffs have been considered!

**Comment by aslilac** [comment]
> I opted for a new endpoint so we don't break dev, and in a followup PR we can remove the old list endpoint

hmm, it's scoped under /provisionerkeys. does it just list built-ins etc. under their "reserved" keys? feels like a bit of a weird place to grab them from, and we can do the grouping on the frontend. I guess this means we only need to make one request tho instead of joining two together.

File: coderd/coderd.go (hunk 6 above)
Comment (f0ssel) [reply coderd/coderd.go:1508]: They should be safe, but I'd rather just follow the no panic principal here so we at least return a proper error in the response

File: enterprise/coderd/provisionerdaemons.go (hunk 7 above)
Comment (aslilac) [reply enterprise/coderd/provisionerdaemons.go]: I guess it's fine to match the health page for now

File: enterprise/coderd/provisionerkeys.go (hunk 8 above)
Comment (Emyrk) [reply enterprise/coderd/provisionerkeys.go]: Reallocating is not a big deal.

File: enterprise/coderd/provisionerdaemons.go (hunk 7 above)
Comment (Emyrk) [reply enterprise/coderd/provisionerdaemons.go]: I don't think this should be a query param. The current stale timer is a hack imo.

In the future, we should just show the current connected provisioner daemons. They all hold open a websocket to 1 Coderd. So we could implement a method of detecting when a provisioner goes away immediately, and not wait for heartbeats.

File: enterprise/coderd/provisionerkeys.go [hunk 9]
```diff
@@ -89,7 +106,7 @@ func (api *API) provisionerKeys(rw http.ResponseWriter, r *http.Request) {
 	ctx := r.Context()
 	organization := httpmw.OrganizationParam(r)
 
-	pks, err := api.Database.ListProvisionerKeysByOrganization(ctx, organization.ID)
+	pks, err := api.Database.ListProvisionerKeysByOrganizationExcludeReserved(ctx, organization.ID)
```
Comment (Emyrk) [inline enterprise/coderd/provisionerkeys.go:109]: Why does the api exclude reserved? Why not include psk and built ins?

**Review by Emyrk — APPROVED** [review]
We might want to make a package that handles all this provisioner and grouping logic.

Not worth it now, but as provisioners are used more within the codebase, it would be annoying as a dev to remember about reserved keys and whatnot.

File: enterprise/coderd/provisionerkeys.go [hunk 10]
```diff
@@ -54,6 +58,19 @@ func (api *API) postProvisionerKey(rw http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	if slices.Contains(codersdk.ReservedProvisionerKeyNames(), req.Name) {
```
Comment (Emyrk) [inline enterprise/coderd/provisionerkeys.go:61]: Might want to make this:
```suggestion
if slices.ContainsFunc(codersdk.ReservedProvisionerKeyNames(), func(s string) bool {
		return strings.EqualFold(req.Name, s)
	}) {
```

To prevent different casings. Would be confusing is all.

**Comment by f0ssel** [comment]
Yeah 💯 I have a feeling this stuff will get changed on the backend at least once before it gets finalized and will do some better grouping then.

File: enterprise/coderd/provisionerkeys.go (hunk 9 above)
Comment (f0ssel) [reply enterprise/coderd/provisionerkeys.go:109]: This endpoint is used by the CLI and I don't think it's necessary to show them in isolation or without a reference to the daemon counterparts since they are more placeholder.

Comment (Emyrk) [reply enterprise/coderd/provisionerkeys.go:109]: Do you think we can deprecate the old endpoint at some point in the future? Should we mark it deprecated in swagger?

File: enterprise/coderd/provisionerdaemons.go (hunk 7 above)
Comment (f0ssel) [reply enterprise/coderd/provisionerdaemons.go]: The realtime data on connected provisioners was a reach goal for GA that we decided was not required, so I think it's totally something we can look at on a future pass.
